content/unroll-default-arguments.md

```diff
@@ -0,0 +1,817 @@
+new parameters can only be added on the right, and they must be provided with a default value.
+
+If there are multiple parameter lists (e.g. for curried methods or methods taking implicits) only one
+parameter list can be unrolled (though it does not need to be the first one). e.g. this works:
+
+```scala
+object Unrolled{
+   def foo(s: String, 
+           n: Int = 1, 
+           @unroll b: Boolean = true, 
+           l: Long = 0)
+          (implicit blah: Blah) = s + n + b + l
+}
+```
+
+As does this
+
+```scala
+object Unrolled{
+   def foo(blah: Blah)
+          (s: String, 
+           n: Int = 1, 
+           @unroll b: Boolean = true, 
+           l: Long = 0) = s + n + b + l
+}
+```
+
+`@unroll`ed methods can be defined in `object`s, `class`es, or `trait`s. Other cases are shown below.
+
+### Unrolling `class`es
+
+Class constructors and secondary constructors are treated by `@unroll` just like any 
+other method:
+
+```scala
+import scala.annotation.unroll
+
+class Unrolled(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0){
+   def foo = s + n + b + l
+}
+```
+
+Unrolls to:
+
+```scala
+import scala.annotation.unroll
+
+class Unrolled(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0){
+   def foo = s + n + b + l
+
+   def this(s: String, n: Int, b: Boolean) = this(s, n, b, 0)
+   def this(s: String, n: Int) = this(s, n, true, 0)
+}
+```
+
+### Unrolling `class` secondary constructors
+
+```scala
+import scala.annotation.unroll
+
+class Unrolled() {
+   var foo = ""
+
+   def this(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0) = {
+      this()
+      foo = s + n + b + l
+   }
+}
+```
+
+Unrolls to:
+
+```scala
+import scala.annotation.unroll
+
+class Unrolled() {
+   var foo = ""
+
+   def this(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0) = {
+      this()
+      foo = s + n + b + l
+   }
+
+   def this(s: String, n: Int, b: Boolean) = this(s, n, b, 0)
+   def this(s: String, n: Int) = this(s, n, true, 0)
+}
+```
+
+### Case Classes
+
+`case class`es can also be `@unroll`ed. Unlike normal `class` constructors
+and method `def`s, `case class`es have several generated methods (`apply`, `copy`)
+that need to be kept in sync with their primary constructor. `@unroll` thus
+generates forwarders for those methods as well, based on the presence of the
+`@unroll` annotation in the primary constructor:
+
+```scala
+import scala.annotation.unroll
+
+case class Unrolled(s: String, n: Int = 1, @unroll b: Boolean = true){
+  def foo = s + n + b
+}
+```
+
+Unrolls to:
+
+```scala
+import scala.annotation.unroll
+
+case class Unrolled(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0L){
+   def this(s: String, n: Int) = this(s, n, true, 0L)
+   def this(s: String, n: Int, b: Boolean) = this(s, n, b, 0L)
+   
+   def copy(s: String, n: Int) = copy(s, n, true, 0L)
+   def copy(s: String, n: Int, b: Boolean) = copy(s, n, b, 0L)
```

Comment:
I wonder if the `copy` can be unrolled like this..., but maybe I'm just confused on how the normal desugaring and default arguments work.

My understanding is that, on a first version of `case class Unrolled(s: String, n: Int = 1)`, the copy method would be declared as `def copy(s: String = this.s, n: Int = this.s)`

Now, say that the user code calls `unrolled.copy(n = 2)`, it will link to that 2 argument version.

Now, on the new version with `case class Unrolled(s: String, n: Int = 1, @unroll b: Boolean = true, l: Long = 0L)`, we will get 3 versions of copy:

- `def copy(s: String = this.s, n: Int = this.s, b: Boolean = this.b, l: Long = this.l)`
- `def copy(s: String, n: Int) = copy(s, n, true, 0L)`
- `def copy(s: String, n: Int, b: Boolean) = copy(s, n, b, 0L)`

Which one will `unrolled.copy(n = 2)` call? Is it able to call the 4 argument version (I imagine not, as that is the point of this SIP, right?)

---

On another note, I'm not sure if `def copy(s: String, n: Int) = copy(s, n, true, 0L)` is correct... Shouldn't it propagate the current values, like `def copy(s: String, n: Int) = copy(s, n, this.b, this.l)`?

Even if the end user is not able to modify those (as they are expecting the old version), the library that is pulling latest version might, and should expect the context to survive under some operations. 🤔

Reply:
> Which one will unrolled.copy(n = 2) call? Is it able to call the 4 argument version (I imagine not, as that is the point of this SIP, right?)

I actually don't know haha. It depends on overload resolution, and I've found it sometimes gives me one and sometimes the other, and I can't figure out why.

Regardless, however, the forwarders and primary implementation *should* have the same semantics, so it's probably fine either way

> On another note, I'm not sure if def copy(s: String, n: Int) = copy(s, n, true, 0L) is correct... Shouldn't it propagate the current values, like def copy(s: String, n: Int) = copy(s, n, this.b, this.l)?

Yes, I think you are correct. It should refer to the current value, not the default value